### PR TITLE
Implemented a trending endpoint

### PR DESCRIPTION
**Pull request by Pchelolo**
Ready for review. The big open question is whether we actually need to store this data.

cc @wikimedia/services @jdlrobson

### CODE REVIEW

**Comment by d00rman** [comment]
I highly doubt that we need to keep historic data for this, honestly. Trending article change during one day as it is, so just proxying and caching ought to be fine.

@jdlrobson @coreyfloyd @berndsi thoughts on historical needs?

File: projects/wmf_default.yaml [hunk 1]
```diff
@@ -63,6 +63,8 @@ paths:
                                 "ttl": 7200}, options.mobileapps)}}'
                 - path: v1/announcements.yaml
                   options: '{{merge({"announcement_cache_control": "s-maxage=86400, max-age=86400"}, options.mobileapps)}}'
+                - path: v1/trending.js
```
Comment (d00rman) [inline projects/wmf_default.yaml]: Hm hm. I wonder if we should now split enwiki off, seeing that trending data is available only for enwiki (for now).

File: lib/mwUtil.js [hunk 2]
```diff
@@ -414,4 +414,86 @@ mwUtil.hydrateResponse = (response, fetch) => {
     .thenReturn(response);
 };
 
+/**
+ * Checks whether the date is today or in the past in UTC-0 timezone
+ *
+ * @param {Date} date a date to check
+ * @return {boolean} true if the date is in the past
+ */
+mwUtil.isHistoric = (date) => {
```
Comment (d00rman) [inline lib/mwUtil.js]: Regardless of whether we will store historical data for the trending edits or not, :+1: for separating this from the feed context.

File: v1/trending.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    getTrending(hyper, req) {
+        mwUtil.verifyDateParams(req);
+        const rp = req.params;
+        const date = mwUtil.getDateSafe(rp);
+        const dateKey = mwUtil.dateToKey(date);
+        const dateArr = dateKey.split('-');
+        const populateSummaries = (res) => {
+            function fetchSummary(uri) {
```
Comment (d00rman) [inline v1/trending.js]: Could this also perhaps be put in `mwUtil` ?

File: v1/trending.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    constructor(options) {
+        this.options = options;
+    }
+
+    _assembleResult(result, dateArr) {
+        // assemble the final response to be returned
+        return {
+            status: 200,
+            headers: {
+                'cache-control': this.options.feed_cache_control,
```
Comment (d00rman) [inline v1/trending.js]: `feed_cache_control`? Why when we have `this.options.cache_control` ?

**Comment by Pchelolo** [comment]
@d00rman Addressed your comments

**Comment by d00rman** [comment]
:+1: Thnx @Pchelolo. I still think it would be prudent to split enwiki's project off of the others, just as we've done with enwiktionary.

File: config.example.wikimedia.yaml [hunk 5]
```diff
@@ -40,6 +40,9 @@ default_project: &default_project
           cache_control: s-maxage=600, max-age=600
           uri: https://pdf-electron.wmflabs.org
           secret: secret
+        trending:
+          host: https://trending-beta.wmflabs.org
+          cache-control: s-maxage=600, max-age=600
```
Comment (d00rman) [inline config.example.wikimedia.yaml]: Let's start conservatively with 30 or 45 mins and then decrease it once we see how the service behaves in production.

**Comment by Pchelolo** [comment]
@d00rman Done - increased caching to 30 minutes and separated out en.wiki

**Comment by d00rman** [comment]
Puppet config change is [here](https://gerrit.wikimedia.org/r/326529).

**Pull request merged**